src/LuaUtils.cpp

```diff
@@ -171,127 +171,363 @@ static bool _import_core(lua_State *L, const std::string &importName)
 
 #undef DEBUG_IMPORT
 #ifdef DEBUG_IMPORT
-#define DEBUG_PRINTF Output
+	#define DEBUG_PRINTF Output
+	#define DEBUG_INDENTED_PRINTF IndentedOutput
+	#define DEBUG_INDENT_INCREASE IndentIncrease
+	#define DEBUG_INDENT_DECREASE IndentDecrease
 #else
-#define DEBUG_PRINTF(...)
+	#define DEBUG_PRINTF(...)
+	#define DEBUG_INDENTED_PRINTF(...)
+	#define DEBUG_INDENT_INCREASE(...)
+	#define DEBUG_INDENT_DECREASE(...)
 #endif
 
-static bool _import(lua_State *L, const std::string &importName)
+// simple struct to pass data between functions
+struct ImportInfo
 {
+	const std::string& importName; // original name argument from "import(importName)"
+	std::string& fileName; // name of the existing file
+	const bool& isFullName; // if true import will no try to load relative to the importDirectories
+	std::vector<std::string>& importDirectories; // contans list of paths where file can be imported
+};
+
+/**
+* Returns path of last element in the lua stack
+*/
+static std::string get_caller(lua_State *L)
+{
+	assert(L);
 	LUA_DEBUG_START(L);
 
-	DEBUG_PRINTF("import: request for %s\n", importName.c_str());
+	std::string caller;
+	lua_Debug ar;
 
-	// get the cache of previously imported things
-	lua_getfield(L, LUA_REGISTRYINDEX, "Imports");
+	// if have stack and caller in the stack
+	if (lua_getstack(L, 1, &ar) && lua_getinfo(L, "S", &ar) && ar.source)
+	{
+		int start = 0;
+		int end = strlen(ar.source);
+
+		if (ar.source[start] != '@' && end > 5)
+			start = 4;
+
+		if (ar.source[start] == '@') {
+			start++;
+
+			static char source[1024];
+			strncpy(source, &ar.source[start], end - start + 1);
+
+			caller = std::string(source);
+		}
+	}
+
+	LUA_DEBUG_END(L, 0);
+	return caller;
+}
+
+/**
+* Finds field in table.
+* Requires a existing table in the stack at index -1.
+* Returns true if field exists and not nil, puts value to lua stack.
+*/
+static bool get_cached(lua_State *L, const std::string& name)
+{
+	LUA_DEBUG_START(L);
 	assert(lua_istable(L, -1));
 
-	// first check the cache for the explicit name (its there if it was
-	// previously imported from core)
-	std::string cacheName = importName;
-	lua_getfield(L, -1, cacheName.c_str());
-	if (!lua_isnil(L, -1)) {
-		// found it, use it
-		DEBUG_PRINTF("import: found cached %s\n", cacheName.c_str());
-		lua_remove(L, -2);
-		return 1;
+	// try to get it
+	lua_getfield(L, -1, name.c_str());
+
+	// if got cached field
+	bool isCached = !lua_isnil(L, -1);
+
+	if (!isCached)
+		lua_pop(L, 1); // remove found "nil" from stack
+
+	LUA_DEBUG_END(L, (isCached ? 1 : 0));
+	return isCached;
+}
+
+/**
+* Finds field in the cache.
+* Require a existing cache table in the stack at index -1.
+* Returns true if field not nil (exists), puts value to lua stack.
+*/
+static bool import_from_cache(lua_State *L, const ImportInfo& importInfo)
+{
+	LUA_DEBUG_START(L);
+	assert(lua_istable(L, -1));
+	DEBUG_INDENTED_PRINTF("import [%s]: trying to load from cache...", importInfo.importName.c_str());
```

Comment:
Don't we have our own variant of printf that understands std::string? (I actually can't remember)

Reply:
@laarmen no, all output functions receives char*. I think it work for other PR =)

Reply:
ok